File: sdk/typescript-client/tests/publication-identity.ts
import type { PublicationIdentity, PublicationRef, InvocationReceipt } from "../src/index.js";
function check(value: unknown): asserts value { if (!value) throw new Error("publication model contract"); }
const component = "sha256:" + "a".repeat(64);
const rows: PublicationIdentity[] = Array.from({length: 4}, (_, i) => ({
  publication: {id: "publication:sha256:" + i.toString(16).padStart(64, "0"), tenant: i < 2 ? "a" : "b"},
  componentDigest: component, packageDigest: "sha256:" + (i % 2).toString(16).padStart(64, "0"),
}));
check(rows[0]!.componentDigest === rows[3]!.componentDigest);
check(rows[0]!.packageDigest === rows[2]!.packageDigest && rows[0]!.packageDigest !== rows[1]!.packageDigest);
check(new Set(rows.map(row => row.publication.id)).size === 4);
const invalid: PublicationRef = {id: "", tenant: "b"};
check(invalid.id === "" && invalid.tenant === "b");
const maximum = (1n << 64n) - 1n;
const unresolved: InvocationReceipt = {activationId: "known", revisionId: "revision", releaseDigest: component,
  routeGeneration: maximum, consumption: {cpuFuel: maximum, peakMemoryBytes: 0n, wallTimeMicros: 0n,
  childCalls: 0, outboundRequests: 0, stateReadBytes: 0n, stateWriteBytes: 0n,
  blobReadBytes: 0n, blobWriteBytes: 0n, logBytes: 0n, effectCount: 0}};
const current: InvocationReceipt = {...unresolved, publicationId: rows[1]!.publication.id};
check(unresolved.publicationId === undefined && current.publicationId === rows[1]!.publication.id);
check(current.releaseDigest === component && current.routeGeneration === maximum && current.consumption.cpuFuel === maximum);

File: sdk/rust/tests/publication_identity.rs
use latent_core::{
    ActivationId, BudgetConsumption, ReleaseDigest, RevisionId, RouteGeneration, TenantId,
};
use latent_sdk::{InvocationReceipt, PublicationIdentity, PublicationRef};

#[test]
fn scoped_coexistence_publication_presence_and_full_width_receipts() {
    let component = ReleaseDigest(format!("sha256:{}", "a".repeat(64)));
    let publications: Vec<_> = (0..4)
        .map(|index| PublicationIdentity {
            publication: PublicationRef {
                id: format!("publication:sha256:{index:064x}").parse().unwrap(),
                tenant: TenantId(if index < 2 { "tenant-a" } else { "tenant-b" }.into()),
            },
            component_digest: component.clone(),
            package_digest: format!("sha256:{:064x}", index % 2).parse().unwrap(),
        })
        .collect();
    assert_eq!(
        publications[0].component_digest,
        publications[3].component_digest
    );
    assert_eq!(
        publications[0].package_digest,
        publications[2].package_digest
    );
    assert_ne!(publications[0].publication, publications[2].publication);
    assert_ne!(
        publications[0].package_digest,
        publications[1].package_digest
    );
    assert!("".parse::<latent_core::PublicationId>().is_err());
    let mut receipt = InvocationReceipt {
        activation_id: ActivationId("known".into()),
        revision_id: RevisionId("revision".into()),
        release_digest: component.clone(),
        publication_id: None,
        route_generation: RouteGeneration(u64::MAX),
        consumption: BudgetConsumption {
            cpu_fuel: u64::MAX,
            ..BudgetConsumption::default()
        },
    };
    assert!(receipt.publication_id.is_none());
    receipt.publication_id = Some(publications[1].publication.id.clone());
    assert_eq!(receipt.release_digest, component);
    assert_eq!(receipt.consumption.cpu_fuel, u64::MAX);
    assert_eq!(receipt.route_generation.0, u64::MAX);
}
